### ISSUE TITLE
How the variable of a custom loss update in distributed training?

### ISSUE THREAD

**Issue by haochange** [facebookresearch/maskrcnn-benchmark#625]
## ❓ Questions and Help
class testLoss(nn.Module):

    def __init__(self):
        super(testLoss, self).__init__()
        self.mat= np.zeros((10, 1024), dtype=np.float32)

    def forward(self, features, index):
       
        features = features.cpu().detach().numpy()
        self.mat[index] = features[index]

If training on one GPU, I'm sure the matrix will be updated correctly. 
If training on multiple gpus, how the variable update? Is there only one matrix in the training? If not, how to do?

**Comment by chengyangfu** [comment]
Hi @haochange, 
   The update is the same. The only potential issue in distributed training is [DDP](https://github.com/facebookresearch/maskrcnn-benchmark/issues/318) which happened when you have unused layers. 
   In distributed training, maskrcnn-benchmark uses the torch.nn.parallel.DistributedDataParalle to synchronize the parameters. If you are interested in this process, check this [file] (https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html). Basically, each GPU contains a copy of the model.  During the backwards pass, gradients from each node are averaged.

**Comment by JoyHuYY1412** [comment]
> Hi @haochange,
> The update is the same. The only potential issue in distributed training is [DDP](https://github.com/facebookresearch/maskrcnn-benchmark/issues/318) which happened when you have unused layers.
> In distributed training, maskrcnn-benchmark uses the torch.nn.parallel.DistributedDataParalle to synchronize the parameters. If you are interested in this process, check this [file] (https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html). Basically, each GPU contains a copy of the model. During the backwards pass, gradients from each node are averaged.

Do you mean here:
https://github.com/facebookresearch/maskrcnn-benchmark/blob/f55844c9aeb5e4d5e6b148c30cd87c64895dd268/maskrcnn_benchmark/engine/trainer.py#L76
the loss backwarding is the loss gathered from all GPUs? How? 
I use multi GPUs on the same machine.
I think this for loop https://github.com/facebookresearch/maskrcnn-benchmark/blob/f55844c9aeb5e4d5e6b148c30cd87c64895dd268/maskrcnn_benchmark/engine/trainer.py#L56
is running in parallel in each GPU, right? If so, are there multi backward pass and multi saving like https://github.com/facebookresearch/maskrcnn-benchmark/blob/f55844c9aeb5e4d5e6b148c30cd87c64895dd268/maskrcnn_benchmark/engine/trainer.py#L104?

Since I only use torch.nn.DataParallel before, I think there is only one process if using torch.nn.DataParallel, right?

**Comment by chengyangfu** [comment]
Hi,  
   This repository uses nn.DistributedDataParallel instead of nn.DataParallel. According to my personal experience, nn.DataParallel has some performance issues related to Python. Maybe PyTorch developers can explain better.
    nn.DistributedDataparallel is like the nn.DataParallel to hide the synchronization details, so the users 
do not need to handle this part by themselves.  Compared to nn.DataParallel, for the machine with 4 gpus as the example, there will be 4 processes and each contains a replicated of the model and a different dataloader. So, each process will run its own forward and backward pass. Then in the forward pass, nn.DistributedDataParallel will synchronize the parameters and buffers (https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html). For the checkpointer, each process all have their own checkpointer, but only the checkpointer in the rank0 process will work
(https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/tools/train_net.py#L49-L52)
